### PR TITLE
Initial RISC-V jit based on IR

### PR DESCRIPTION
**Pull request by unknownbrackets**
~This isn't quite ready, as it's still slower than interpreter (mostly during loading though, probably needs load/store implemented.)~

This is now often faster than the interpreter, although not so much when floats/vecs are heavily used.

#17771, #17773, and #17776 should be merged first since this enables jit by default.

Need to test more games, but so far most run pretty well (barring some IR bugs.)  Tests also pass (except 2 that also fail for IR.)

Flush to zero looks like it will be a problem (and I recall Gods Eater Burst requires it for save validation...)  RISC-V doesn't seem to intend to expose it.

Notes on the implementation:
 * Currently based on IR heavily, I wanted to see the pros/cons of using its block cache which is pretty different (also gives me preload which may help me find bugs easier.)  Probably makes sense to switch to a "full fledged" backend using the regular block cache.
 * A decent number of ALU related ops implemented now, and does have static alloc etc.  No floats yet.
 * Crash handler is not well tested (but seems to work.)
 * Rounding hasn't been thought through much.
 * Periodic trampolines or constant polls may make sense for far jumps or references to C++ globals.
 * SCRATCH1/SCRATCH2 should probably not be a0/a1, worried it'll cause confusion in func calls (save/load regs etc.)  But they are regs that can be used in 16-bit instructions...

-[Unknown]

### CODE REVIEW

**Comment by hrydgard** [comment]
Very cool to try IR -> native! Will look at this soon, and looking forward to following this. 

What do you mean by periodic trampolines?

**Comment by unknownbrackets** [comment]
RISC-V can only jump +/- 1MB, so even going back to the dispatcher, rounding, save/load regs, block linking, etc. requires a register and jump.  It does have AUIPC (which the emitter is currently not optimally using) which allows a +/-31 bit jump in two instructions, so maybe that's enough for semi-local jumps.

But anywhere we're calling (or loading) things outside the jit space, constructing a 64-bit immediate can take 6 instructions (with a second temporary reg) and for example clang uses literal pools for 64-bit constants instead.  So a periodic trampoline/pool might make sense, or might not - if only from a codesize perspective.

-[Unknown]

**Comment by hrydgard** [comment]
We might just want to make more effort to squeeze everything together into a 31-bit range then, so we can use that 2-instruction combo. Not sure how feasible though.

**Comment by unknownbrackets** [comment]
Well, non-jit code (by which I mean, Advance, syscalls, coreState, etc.) won't always be there.  It's similiar to the problem on x86 with RIP addressing.

-[Unknown]

**Comment by hrydgard** [comment]
Well - they aren't in range currently, but they could be made to be, with careful enough usage of mmap etc. We could suballocate all our executable buffers from one larger one for example and even have that suballocated from a larger block that we manually put stuff like coreState in, though it will be some extra complexity.

File: Core/MIPS/RiscV/RiscVRegCache.cpp [hunk 1]
```diff
@@ -0,0 +1,1075 @@
...
+	_dbg_assert_(IsValidReg(mipsReg));
+	_dbg_assert_(!mr[mipsReg].isStatic);
+	if (mr[mipsReg].isStatic) {
+		ERROR_LOG(JIT, "Cannot MapRegTo static register %d", mipsReg);
+		return;
+	}
+	ar[reg].isDirty = (mapFlags & MIPSMap::DIRTY) == MIPSMap::DIRTY;
+	if ((mapFlags & MIPSMap::NOINIT) != MIPSMap::NOINIT) {
+		if (mipsReg == MIPS_REG_ZERO) {
+			// If we get a request to load the zero register, at least we won't spend
```
Comment (hrydgard) [inline Core/MIPS/RiscV/RiscVRegCache.cpp:289]: in the comment: s/load/map

**Review by hrydgard — APPROVED** [review]
Again, really good to see a first stab at a IR->native JIT, thanks for doing this!

I had somehow hoped for the post-IR regcache to get simpler, turns out it's not really, for good reasons. At least there's less of a requirement for the instruction emulations to collapse immediates since that's already done by the IR optimizer, which is nice.

Overall, this looks great. Just some nit-picking and some comments that don't need addressing.

I'm surprised at the lack of denormal flushing support in the ISA, that's gonna be an issue for things like Proton and other emulators too...

File: Core/MIPS/RiscV/RiscVJit.cpp [hunk 2]
```diff
@@ -0,0 +1,557 @@
...
+void RiscVJit::NormalizeSrc1(IRInst inst, RiscVReg *reg, RiscVReg tempReg, bool allowOverlap) {
+	*reg = NormalizeR(inst.src1, allowOverlap ? 0 : inst.dest, tempReg);
+}
+
+void RiscVJit::NormalizeSrc12(IRInst inst, RiscVReg *lhs, RiscVReg *rhs, RiscVReg lhsTempReg, RiscVReg rhsTempReg, bool allowOverlap) {
+	*lhs = NormalizeR(inst.src1, allowOverlap ? 0 : inst.dest, lhsTempReg);
+	*rhs = NormalizeR(inst.src2, allowOverlap ? 0 : inst.dest, rhsTempReg);
+}
+
+RiscVReg RiscVJit::NormalizeR(IRRegIndex rs, IRRegIndex rd, RiscVReg tempReg) {
```
[the remark this answers is not in the document]
Comment (unknownbrackets) [reply Core/MIPS/RiscV/RiscVJit.cpp:545]: We can't.  There's no ANDW, and there's also no BGEW or SLTIW.  Those don't exist, so sometimes you need to normalize.

-[Unknown]

[the remark this answers is not in the document]
Comment (hrydgard) [reply Core/MIPS/RiscV/RiscVJit.cpp:545]: Ah, I didn't realize. I guess that's the end of that idea, then.

File: Core/MIPS/RiscV/RiscVJit.cpp [hunk 3]
```diff
@@ -0,0 +1,557 @@
...
+			//fpr.FlushAll();
+		}
+
+		// Safety check, in case we get a bunch of really large jit ops without a lot of branching.
+		if (GetSpaceLeft() < 0x800) {
+			return false;
+		}
+	}
+
+	// Note: a properly constructed block should never get here.
```
[the remark this answers is not in the document]
Comment (unknownbrackets) [reply Core/MIPS/RiscV/RiscVJit.cpp:101]: Well, the loop can get here just fine, it's just that jit execution shouldn't.  I could QuickCallFunction to a log/assert.

-[Unknown]

File: Core/MIPS/RiscV/RiscVJit.cpp [hunk 4]
```diff
@@ -0,0 +1,557 @@
...
+	default:
+		_assert_msg_(false, "Unexpected IR op %d", (int)inst.op);
+		CompIR_Generic(inst);
+		break;
+	}
+}
+
+void RiscVJit::CompIR_Generic(IRInst inst) {
+	// For now, we're gonna do it the slow and ugly way.
+	// Maybe there's a smarter way to fallback?
```
[the remark this answers is not in the document]
Comment (unknownbrackets) [reply Core/MIPS/RiscV/RiscVJit.cpp:397]: I'll make it check the exit flag on IRMeta to at least skip the return value check.

-[Unknown]

File: Core/MIPS/RiscV/RiscVJit.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (hrydgard) [reply Core/MIPS/RiscV/RiscVJit.cpp:101]: Oh, right. Well, might not be worth the trouble.

**Pull request merged**